File: s3.go
package main

import (
	"context"
	"fmt"
	"net/http"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/credentials"
	"github.com/aws/aws-sdk-go-v2/service/s3"
	"github.com/gin-gonic/gin"
)

type S3Service struct {
	client    *s3.Client
	presigner *s3.PresignClient
	bucket    string
	region    string
}

func NewS3Service(ctx context.Context, bucket, region, endpoint, accessKey, secretKey string) (*S3Service, error) {
	opts := []func(*config.LoadOptions) error{
		config.WithRegion(region),
	}

	if endpoint != "" {
		customResolver := aws.EndpointResolverWithOptionsFunc(
			func(service, region string, options ...interface{}) (aws.Endpoint, error) {
				return aws.Endpoint{URL: endpoint}, nil
			},
		)
		opts = append(opts, config.WithEndpointResolverWithOptions(customResolver))
	}

	if accessKey != "" && secretKey != "" {
		opts = append(opts, config.WithCredentialsProvider(
			credentials.NewStaticCredentialsProvider(accessKey, secretKey, ""),
		))
	}

	cfg, err := config.LoadDefaultConfig(ctx, opts...)
	if err != nil {
		return nil, fmt.Errorf("failed to load AWS config: %w", err)
	}

	client := s3.NewFromConfig(cfg)
	presigner := s3.NewPresignClient(client)

	return &S3Service{
		client:    client,
		presigner: presigner,
		bucket:    bucket,
		region:    region,
	}, nil
}

func (s *S3Service) GenerateUploadURL(ctx context.Context, convID, fileExt string) (string, string, error) {
	fileKey := GenerateFileKey(convID, fileExt)

	result, err := s.presigner.PresignPutObject(ctx, &s3.PutObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(fileKey),
	}, func(opts *s3.PresignOptions) {
		opts.Expires = 2 * time.Minute
	})

	if err != nil {
		return "", "", fmt.Errorf("failed to generate upload URL: %w", err)
	}

	return result.URL, fileKey, nil
}

func (s *S3Service) GenerateDownloadURL(ctx context.Context, fileKey string) (string, error) {
	result, err := s.presigner.PresignGetObject(ctx, &s3.GetObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(fileKey),
	}, func(opts *s3.PresignOptions) {
		opts.Expires = 7 * 24 * time.Hour
	})

	if err != nil {
		return "", fmt.Errorf("failed to generate download URL: %w", err)
	}

	return result.URL, nil
}

func (s *S3Service) DeleteObject(ctx context.Context, fileKey string) error {
	_, err := s.client.DeleteObject(ctx, &s3.DeleteObjectInput{
		Bucket: aws.String(s.bucket),
		Key:    aws.String(fileKey),
	})
	if err != nil {
		return fmt.Errorf("failed to delete object: %w", err)
	}
	return nil
}

func GenerateFileKey(convID, fileExt string) string {
	id := GenerateNanoID(21)
	return fmt.Sprintf("uploads/%s/%s.%s", convID, id, fileExt)
}

func ExtractConvIDFromKey(fileKey string) (string, error) {
	parts := strings.Split(fileKey, "/")
	if len(parts) < 3 || parts[0] != "uploads" {
		return "", fmt.Errorf("invalid file key format")
	}
	return parts[1], nil
}

type PresignRequest struct {
	Operation string `json:"operation" binding:"required,oneof=upload download"`
	ConvID    string `json:"conv_id"`
	FileExt   string `json:"file_ext"`
	FileKey   string `json:"file_key"`
}

// GetPresignURL 获取文件预签名 URL
// @Summary      获取文件预签名 URL
// @Description  获取文件上传或下载的预签名 URL
// @Tags         files
// @Accept       json
// @Produce      json
// @Security     bearerAuth
// @Param        request  body  PresignRequest  true  "预签名请求"
// @Success      200  {object}  map[string]interface{}  "预签名 URL"
// @Failure      400  {object}  ErrorResponse  "请求无效"
// @Failure      501  {object}  ErrorResponse  "S3 服务未配置"
// @Router       /api/files/presign [post]
func (h *Handler) GetPresignURL(c *gin.Context) {
	userID := c.GetInt64("user_id")

	if h.svc.s3Service == nil {
		c.JSON(http.StatusNotImplemented, gin.H{"error": "S3 service not configured"})
		return
	}

	var req PresignRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		handleError(c, ErrInvalidInput)
		return
	}

	switch req.Operation {
	case "upload":
		if req.ConvID == "" || req.FileExt == "" {
			handleError(c, ErrInvalidInput)
			return
		}

		if err := h.checkConvPermission(c.Request.Context(), req.ConvID, userID); err != nil {
			handleError(c, err)
			return
		}

		url, fileKey, err := h.svc.s3Service.GenerateUploadURL(c.Request.Context(), req.ConvID, req.FileExt)
		if err != nil {
			handleError(c, err)
			return
		}

		c.JSON(http.StatusOK, gin.H{
			"url":        url,
			"file_key":   fileKey,
			"method":     "PUT",
			"expires_in": 120,
		})

	case "download":
		if req.FileKey == "" {
			handleError(c, ErrInvalidInput)
			return
		}

		convID, err := ExtractConvIDFromKey(req.FileKey)
		if err != nil {
			handleError(c, ErrInvalidInput)
			return
		}

		if err := h.checkConvPermission(c.Request.Context(), convID, userID); err != nil {
			handleError(c, err)
			return
		}

		url, err := h.svc.s3Service.GenerateDownloadURL(c.Request.Context(), req.FileKey)
		if err != nil {
			handleError(c, err)
			return
		}

		c.JSON(http.StatusOK, gin.H{
			"url":        url,
			"file_key":   req.FileKey,
			"method":     "GET",
			"expires_in": 604800,
		})
	}
}

// checkConvPermission 检查用户是否有权限访问指定会话（上传/下载通用）
func (h *Handler) checkConvPermission(ctx context.Context, convID string, userID int64) error {
	hasPerm, err := h.svc.convPartStore.Exists(ctx, h.svc.convPartStore.DB(), convID, userID)
	if err != nil {
		return err
	}
	if !hasPerm {
		if groupID := ExtractGroupID(convID); groupID != "" {
			isMember, _ := h.svc.IsUserInGroup(ctx, groupID, userID)
			if !isMember {
				return ErrForbidden
			}
		} else {
			return ErrForbidden
		}
	}
	return nil
}

File: src/s3.rs
use aws_config::BehaviorVersion;
use aws_sdk_s3::Client;
use aws_sdk_s3::config::Credentials;
use aws_sdk_s3::presigning::PresigningConfig;
use axum::Json;
use axum::extract::State;
use axum::http::StatusCode;
use axum::response::IntoResponse;
use serde::{Deserialize, Serialize};
use std::time::Duration;

use crate::error::AppError;
use crate::shared::Service;

pub struct S3Service {
    client: Client,
    presigner: Client,
    bucket: String,
}

impl Clone for S3Service {
    fn clone(&self) -> Self {
        Self {
            client: self.client.clone(),
            presigner: self.presigner.clone(),
            bucket: self.bucket.clone(),
        }
    }
}

impl S3Service {
    pub async fn new(
        bucket: &str,
        region: &str,
        endpoint: &str,
        access_key: &str,
        secret_key: &str,
    ) -> Result<Self, String> {
        let region = aws_sdk_s3::config::Region::new(region.to_string());
        let mut config_builder = aws_config::defaults(BehaviorVersion::latest()).region(region);

        if !endpoint.is_empty() {
            config_builder = config_builder.endpoint_url(endpoint);
        }

        if !access_key.is_empty() && !secret_key.is_empty() {
            let creds = Credentials::new(access_key, secret_key, None, None, "static");
            config_builder = config_builder.credentials_provider(creds);
        }

        let sdk_config = config_builder.load().await;
        let client = Client::new(&sdk_config);

        Ok(Self {
            client: client.clone(),
            presigner: client,
            bucket: bucket.to_string(),
        })
    }

    pub async fn generate_upload_url(
        &self,
        conv_id: &str,
        file_ext: &str,
    ) -> Result<(String, String), AppError> {
        let file_key = generate_file_key(conv_id, file_ext);

        let presign_config = PresigningConfig::builder()
            .expires_in(Duration::from_secs(120))
            .build()
            .map_err(|e| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

        let presigned = self
            .presigner
            .put_object()
            .bucket(&self.bucket)
            .key(&file_key)
            .presigned(presign_config)
            .await
            .map_err(|e| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

        Ok((presigned.uri().to_string(), file_key))
    }

    pub async fn generate_download_url(&self, file_key: &str) -> Result<String, AppError> {
        let presign_config = PresigningConfig::builder()
            .expires_in(Duration::from_secs(604800))
            .build()
            .map_err(|e| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

        let presigned = self
            .presigner
            .get_object()
            .bucket(&self.bucket)
            .key(file_key)
            .presigned(presign_config)
            .await
            .map_err(|e| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

        Ok(presigned.uri().to_string())
    }

    pub async fn delete_object(&self, file_key: &str) -> Result<(), AppError> {
        self.client
            .delete_object()
            .bucket(&self.bucket)
            .key(file_key)
            .send()
            .await
            .map_err(|e| AppError::new(StatusCode::INTERNAL_SERVER_ERROR, &e.to_string()))?;

        Ok(())
    }
}

#[derive(Debug, Deserialize)]
pub struct PresignRequest {
    pub operation: String,
    #[serde(default)]
    pub conv_id: String,
    #[serde(default)]
    pub file_ext: String,
    #[serde(default)]
    pub file_key: String,
}

#[derive(Debug, Serialize)]
pub struct PresignResponse {
    pub url: String,
    pub file_key: String,
    pub method: String,
    pub expires_in: u64,
}

impl Service {
    pub async fn get_presign_url(
        &self,
        user_id: i64,
        req: PresignRequest,
    ) -> Result<impl IntoResponse + use<>, AppError> {
        let s3_service = self.s3_service.as_ref().ok_or_else(|| {
            AppError::new(
                StatusCode::INTERNAL_SERVER_ERROR,
                "S3 service not configured",
            )
        })?;

        match req.operation.as_str() {
            "upload" => {
                if req.conv_id.is_empty() || req.file_ext.is_empty() {
                    return Err(AppError::bad_request("conv_id and file_ext are required"));
                }

                let has_perm = self.conv_part_store.exists(&req.conv_id, user_id).await?;

                if !has_perm {
                    if let Some(group_id) = crate::shared::extract_group_id(&req.conv_id) {
                        let is_member = self.group_mem_store.is_member(&group_id, user_id).await?;
                        if !is_member {
                            return Err(AppError::forbidden());
                        }
                    } else {
                        return Err(AppError::forbidden());
                    }
                }

                let (url, file_key) = s3_service
                    .generate_upload_url(&req.conv_id, &req.file_ext)
                    .await?;

                Ok(Json(PresignResponse {
                    url,
                    file_key,
                    method: "PUT".to_string(),
                    expires_in: 120,
                }))
            }
            "download" => {
                if req.file_key.is_empty() {
                    return Err(AppError::bad_request("file_key is required"));
                }

                let conv_id = extract_conv_id_from_key(&req.file_key)
                    .ok_or_else(|| AppError::bad_request("invalid file key format"))?;

                let has_perm = self.conv_part_store.exists(&conv_id, user_id).await?;

                if !has_perm {
                    if let Some(group_id) = crate::shared::extract_group_id(&conv_id) {
                        let is_member = self.group_mem_store.is_member(&group_id, user_id).await?;
                        if !is_member {
                            return Err(AppError::forbidden());
                        }
                    } else {
                        return Err(AppError::forbidden());
                    }
                }

                let url = s3_service.generate_download_url(&req.file_key).await?;

                Ok(Json(PresignResponse {
                    url,
                    file_key: req.file_key,
                    method: "GET".to_string(),
                    expires_in: 604800,
                }))
            }
            _ => Err(AppError::bad_request("invalid operation")),
        }
    }
}

pub async fn presign(
    State(state): State<crate::AppState>,
    axum::extract::Extension(user_id): axum::extract::Extension<i64>,
    Json(req): Json<PresignRequest>,
) -> Result<impl IntoResponse + use<>, AppError> {
    state.svc.get_presign_url(user_id, req).await
}

pub fn generate_file_key(conv_id: &str, file_ext: &str) -> String {
    let id = uuid::Uuid::new_v4();
    format!("uploads/{}/{}.{}", conv_id, id, file_ext)
}

pub fn extract_conv_id_from_key(file_key: &str) -> Option<String> {
    let parts: Vec<&str> = file_key.split('/').collect();
    if parts.len() >= 3 && parts[0] == "uploads" {
        Some(parts[1].to_string())
    } else {
        None
    }
}
